File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        generate("", n, 0, 0, result);
        return result;
    }
    
    void generate(string current, int n, int open, int close, vector<string>& result) {
        if (open == n && close == n) {
            result.push_back(current);
            return;
        }
        
        if (open < n) {
            generate(current + "(", n, open+1, close, result);
        }
        
        if (close < open) {
            generate(current + ")", n, open, close+1, result);
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const result = [];
    generateParenthesisHelper(result, "", 0, 0, n);
    return result;
};

function generateParenthesisHelper(result, current, open, close, n) {
    if (current.length === n * 2) {
        result.push(current);
        return;
    }

    if (open < n) {
        generateParenthesisHelper(result, current + "(", open + 1, close, n);
    }

    if (close < open) {
        generateParenthesisHelper(result, current + ")", open, close + 1, n);
    }
}
